// balanced-match@2.0.0-master/index.js
'use strict';

// 모듈 내보내기 함수 정의
module.exports = balanced;

function balanced(a, b, str) {
  // 만약 a가 정규 표현식이라면, 문자열에서 해당 정규 표현식과 일치하는 부분을 찾음
  if (a instanceof RegExp) a = maybeMatch(a, str);
  // 만약 b가 정규 표현식이라면, 문자열에서 해당 정규 표현식과 일치하는 부분을 찾음
  if (b instanceof RegExp) b = maybeMatch(b, str);

  // 범위 계산 함수를 사용하여 균형잡힌 문자열 범위를 찾음
  var r = range(a, b, str);

  // 찾은 범위가 있으면 해당 범위를 반환함
  return r && {
    start: r[0],
    end: r[1],
    pre: str.slice(0, r[0]),
    body: str.slice(r[0] + a.length, r[1]),
    post: str.slice(r[1] + b.length)
  };
}

// 정규 표현식과 문자열이 매치되는지 확인하고 매치된 부분을 반환하는 함수
function maybeMatch(reg, str) {
  var m = str.match(reg);
  return m ? m[0] : null;
}

// 범위 계산 함수 내보내기
balanced.range = range;

// 균형잡힌 문자열 범위를 계산하는 함수 정의
function range(a, b, str) {
  var begs, beg, left, right, result;
  var ai = str.indexOf(a);
  var bi = str.indexOf(b, ai + 1);
  var i = ai;

  // a와 b가 문자열 내에 모두 존재하는 경우
  if (ai >= 0 && bi > 0) {
    // a와 b가 같은 경우
    if(a===b) {
      return [ai, bi];
    }
    begs = [];
    left = str.length;

    // 문자열을 순회하면서 균형잡힌 범위를 찾음
    while (i >= 0 && !result) {
      if (i == ai) {
        begs.push(i);
        ai = str.indexOf(a, i + 1);
      } else if (begs.length == 1) {
        result = [ begs.pop(), bi ];
      } else {
        beg = begs.pop();
        if (beg < left) {
          left = beg;
          right = bi;
        }

        bi = str.indexOf(b, i + 1);
      }

      i = ai < bi && ai >= 0 ? ai : bi;
    }

    if (begs.length) {
      result = [ left, right ];
    }
  }

  return result;
}
